packages/yarnpkg-fslib/sources/ZipFS.ts

```diff
@@ -1204,21 +1256,35 @@ export class ZipFS extends BasePortableFakeFS {
     if (typeof index === `undefined`)
       throw errors.EINVAL(`rmdir '${p}'`);
 
-    this.libzip.delete(this.zip, index);
     this.unregisterEntry(resolvedP);
+
+    const rc = this.libzip.delete(this.zip, index);
+    if (rc === -1) {
+      throw this.makeLibzipError(this.libzip.getError(this.zip));
+    }
   }
 
   private hydrateDirectory(resolvedP: PortablePath) {
     const index = this.libzip.dir.add(this.zip, ppath.relative(PortablePath.root, resolvedP));
     if (index === -1)
-      throw new Error(this.libzip.error.strerror(this.libzip.getError(this.zip)));
+      throw this.makeLibzipError(this.libzip.getError(this.zip));
 
     this.registerListing(resolvedP);
     this.registerEntry(resolvedP, index);
 
     return index;
   }
 
+  async linkPromise(existingP: PortablePath, newP: PortablePath) {
+    return this.linkSync(existingP, newP);
+  }
+
+  linkSync(existingP: PortablePath, newP: PortablePath) {
+    // Zip archives don't support hard links:
+    // https://stackoverflow.com/questions/8859616/are-hard-links-possible-within-a-zip-archive
+    throw errors.EOPNOTSUPP(`link '${existingP}' -> '${newP}'`);
```

Comment:
I used `ENOSYS` for other "not implemented" errors, is there a reference for when to use `EOPNOTSUPP`?

Reply:
`link` is not implemented in `ZipFS` because Zip archives don't support hardlinks. I'd consider this (after reading https://www.gnu.org/software/libc/manual/html_node/Error-Codes.html) more as an `Operation not supported` (`EOPNOTSUPP`), rather than as a `Function not implemented` (`ENOSYS`).

I think that:
- we should keep `EOPNOTSUPP` if we decide that the behavior of hardlinks doesn't make sense for `ZipFS`
- we should use `ENOSYS` if we decide that we'll implement mocked hardlinks in `ZipFS` in the future (they would have the same behavior as hardlinks regarding `stat.ino`, `stat.nlink` and `setFileSource`, but each entry would have take up space on disk, since we can't make them share the same space)

What do you think?

Reply:
I agree with this semantic 👍 Not sure about hardlinks support though - they are very different, as changing one won't change the other. A code expecting a hardlink to work could get really weird bugs 🤔

Reply:
I'd say that we should keep it the way it currently is until we decide what to do with hardlinks.